Add a encryption option to support previous data encrypted non-deterministically with a SHA1 hash digest

jorgemanrubia opened the pull request:
This adds a new Active Record encryption option to support previous data encrypted non-deterministically with a SHA1 hash digest.

There is currently a problem with Active Record encryption for users updating from 7.0 to 7.1 Before #44873, data encrypted with non-deterministic encryption was always using SHA-1. The reason is that `ActiveSupport::KeyGenerator.hash_digest_class` was set in an `after_initialize` block in the railtie config, but encryption config was running before that, so it was effectively using the previous default SHA1. That means that existing users are using SHA256 for non deterministic encryption, and SHA1 for deterministic encryption.

This adds a new option `support_sha1_for_non_deterministic_encryption` that users can enable to support for SHA1 and SHA256 when decrypting existing data.

The new option will kick-in by default for versions prior to 7.1. The idea is that:

- Existing users updating from 7.0, won't have to set it, as it will get enabled by default.
- Users starting with 7.1 will have the option disabled.

This also improves the initialization system for AR encryption to make it more robust. It will now happen after all the initializers have run, and the system to declare filtered attributes will work with different load orders.

Big thanks to @duduribeiro for the thorough investigation and debugging here, as he was the one identifying the issue and who explored alternatives to fix the problem (see #48520). I added him as the co-author of the commit I ported from that PR.

Pending: 

- [x] Validate patch in real scenarios
- [x] Update new framework default initializer
- [x] Update CHANGELOG
- [x] Update documentation
- [x] Suite passing

Fixes https://github.com/rails/rails/issues/48204. 
Closes https://github.com/rails/rails/pull/48520.

jorgemanrubia committed:
Make sure active record encryption configuration happens after initializers have run

Co-authored-by: Cadu Ribeiro <mail@cadu.dev>

jorgemanrubia committed:
Add a new option to support previous data encrypted non-deterministically with a hash digest of SHA1

There is currently a problem with Active Record encryption for users updating from 7.0 to 7.1 Before
#44873, data encrypted with non-deterministic encryption was always using SHA-1. The reason is that
`ActiveSupport::KeyGenerator.hash_digest_class` is set in an after_initialize block in the railtie config,
but encryption config was running before that, so it was effectively using the previous default SHA1. That
means that existing users are using SHA256 for non deterministic encryption, and SHA1 for deterministic
encryption.

This adds a new option `use_sha1_digest_for_non_deterministic_data` that
users can enable to support for SHA1 and SHA256 when decrypting existing data.

eileencodes commented on activerecord/lib/active_record/encryption/config.rb:
```diff
@@ -42,6 +51,7 @@ def set_defaults
           self.previous_schemes = []
           self.forced_encoding_for_deterministic_encryption = Encoding::UTF_8
           self.hash_digest_class = OpenSSL::Digest::SHA1
+          self.support_previous_data_encrypted_non_deterministically_with_sha1_class_digest = false
```
This option name feels like quite the mouthful, I think we can shorten it. How about something like `use_sha1_digest_for_non_deterministic_data`?

jorgemanrubia replied:
I'll go with `support_sha1_digest_for_non_deterministic_data` since it will still use sha-256 at encryption time, and support both at decryption time.

jorgemanrubia committed:
Set a default value of true for `support_sha1_for_non_deterministic_encryption` and proper initializer values.

We want to enable the flag existing versions (< 7.1), and we want it to be false moving by
default moving forward.

jorgemanrubia commented:
> As this might happen again in the future when someone might decide to use something other than SHA265 or use another private key, should Active Record Encryption support rotation?

Yes @p8. The library supports rotations via support for previous encryption schemes, and the PR is actually leveraging that, but this should be more streamlined and better documented.

jorgemanrubia commented:
@duduribeiro I'd appreciate if you can check if this PR is working for the failing scenario you had. When upgrading from 7.0, you should add this to the framework version initializer:

```
Rails.application.config.active_record.encryption.hash_digest_class = OpenSSL::Digest::SHA256
```

p8 commented:
> > As this might happen again in the future when someone might decide to use something other than SHA265 or use another private key, should Active Record Encryption support rotation?
> 
> Yes @p8. The library supports rotations via support for previous encryption schemes, and the PR is actually leveraging that, but this should be more streamlined and better documented.

Good to know @jorgemanrubia. Sorry, I missed rotations are already supported.

duduribeiro commented:
@jorgemanrubia but this is maybe something on my side on my app. Checking more 😄 

On my test app for this bug https://github.com/duduribeiro/testbug it worked just fine!!! (ive pushed the commit to use this patch)

thanks so much for it!

duduribeiro commented:
@jorgemanrubia I will investigate this and will thrown my findings here. 

One thing that I noticed is that when I boot the app I have the same behave I fixed on https://github.com/rails/rails/pull/48520

```ruby
Loading development environment (Rails 7.1.0.alpha)
irb(main):001:0> ActiveRecord::Encryption.config.hash_digest_class
=> OpenSSL::Digest::SHA1
irb(main):002:0> 
```

even set `Rails.application.config.active_record.encryption.hash_digest_class = OpenSSL::Digest::SHA256` on `config/initializers/new_framework...`

duduribeiro commented:
also noticed this when booting the app:

```ruby
[22, 31] in ~/.asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/bundler/gems/rails-3bdccc24a416/activerecord/lib/active_record/encryption/configurable.rb
    22|           config.deterministic_key = deterministic_key
    23|           config.key_derivation_salt = key_derivation_salt
    24| 
    25|           # Set the default for this property here instead of in +Config#set_defaults+ as this needs
    26|           # to happen *after* the keys have been set.
=>  27|           debugger
    28|           properties[:support_sha1_for_non_deterministic_encryption] = true if properties[:support_sha1_for_non_deterministic_encryption].nil?
    29| 
    30|           properties.each do |name, value|
    31|             ActiveRecord::Encryption.config.send "#{name}=", value if ActiveRecord::Encryption.config.respond_to?("#{name}=")
=>#0	ActiveRecord::Encryption::Configurable::ClassMethods#configure(primary_key="usvSIvN1fPKSTeiSoyxJBcb3PWUB4Klv", deterministic_key="R82uMg7XnziPOGIfQsTmTOsHPbqkep29", key_derivation_salt="rOmFmA7vvd88oUyF635MXUyPhkXA1uYR", properties={}) at ~/.asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/bundler/gems/rails-3bdccc24a416/activerecord/lib/active_record/encryption/configurable.rb:27
  #1	block in <class:Railtie> (2 levels) at ~/.asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/bundler/gems/rails-3bdccc24a416/activerecord/lib/active_record/railtie.rb:363
  # and 43 frames (use `bt' command for all frames)
(ruby) properties[:support_sha1_for_non_deterministic_encryption]
nil
(rdbg) properties
{}
```

duduribeiro commented:
@jorgemanrubia now is happening something.. weird 😄 

name is a non-deterministic. looks like it worked fine (although I think it is always setting it as `true` the `support_sha1_for_non_deterministic_encryption`, because when I set `Rails.application.config.active_record.encryption.support_sha1_for_non_deterministic_encryption` to `false` it does not raises any error)

but for `document` ( the deterministic one), it is printing the `raw` value stored on the column.
```ruby
irb(main):004:0> Person.last.name 
  Person Load (6.0ms)  SELECT "people".* FROM "people" ORDER BY "people"."id" DESC LIMIT $1  [["LIMIT", 1]]
=> "Carlos Eduardo Ribeiro"
irb(main):005:0> Person.last.document
  Person Load (3.4ms)  SELECT "people".* FROM "people" ORDER BY "people"."id" DESC LIMIT $1  [["LIMIT", 1]]
=> "{\"p\":\"avxyZPZN+eSuB4g=\",\"h\":{\"iv\":\"8k1KLwy582tX4wQb\",\"at\":\"u2f4TGaLw5zWRgz17OTsvg==\"}}"
```

duduribeiro commented:
@jorgemanrubia Ok. I think it is an initialization problem again:

when I set the digest_class on `application.rb` it works just fine:

```ruby
cadu@MacBook-Pro-de-Cadu adaflow % cat config/application.rb 
require_relative "boot"

require "rails/all"

# Require the gems listed in Gemfile, including any gems
# you've limited to :test, :development, or :production.
Bundler.require(*Rails.groups)

require "jumpstart"

module Myapp
  class Application < Rails::Application
    # Initialize configuration defaults for originally generated Rails version.
    config.load_defaults 7.0

    # Settings in config/environments/* take precedence over those specified here.
    # Application configuration can go into files in config/initializers
    # -- all .rb files in that directory are automatically loaded after loading
    # the framework and any gems in your application.

    config.exceptions_app = routes

    # Where the I18n library should search for translation files
    # Search nested folders in config/locales for better organization
    config.i18n.load_path += Dir[Rails.root.join("config", "locales", "**", "*.{rb,yml}")]

    # Permitted locales available for the application
    config.i18n.available_locales = [:en, :"pt-BR"]

    # Set default locale
    config.i18n.default_locale = :"pt-BR" unless Rails.env.test?

    # Use default language as fallback if translation is missing
    config.i18n.fallbacks = true

    # Prevent sassc-rails from setting sass as the compressor
    # Libsass is deprecated and doesn't support modern CSS syntax used by TailwindCSS
    config.assets.css_compressor = nil

    # Rails 7 defaults to libvips as the variant processor
    # libvips is up to 10x faster and consumes 1/10th the memory of imagemagick
    # If you need to use imagemagick, uncomment this to switch
    # config.active_storage.variant_processor = :mini_magick

    config.active_record.encryption.hash_digest_class = OpenSSL::Digest::SHA256 <<<<<<<
  end
end
```

jorgemanrubia commented:
> yes. it fixed it! are u going to make a patch to have an after_initialize on the new code or is this some problem that I have on my app that shouldn't be occuring and I need to fix it?

Oh great. Yeah I'll change it as part of this PR. I think it's safer, since it will guarantee it runs after the initializers despite of how active record loads, which is exactly what we want. Thanks!

duduribeiro commented:
@jorgemanrubia just for curiosity. I found the culprit of loading it early than initializers 😄 

I use https://github.com/basecamp/name_of_person on my app and when I do 
```ruby
gem "name_of_person", "~> 1.1"
```

on my app it loads `active_record_encryption.configuration` before the initializers have run. 

Ive updated https://github.com/duduribeiro/testbug to add the name_of_person gem so we can test with new fix

jorgemanrubia commented:
Nice catch @duduribeiro. That's good evidence that using `after_initializer` is a more robust approach here. I pushed a fix, and I'm waiting for the suite to pass. I'd appreciate if you can give it a test with you project.

duduribeiro commented:
> Nice catch @duduribeiro. That's good evidence that using `after_initializer` is a more robust approach here. I pushed a fix, and I'm waiting for the suite to pass. I'd appreciate if you can give it a test with you project.

@jorgemanrubia tested on my app and it worked perfectly 🎉. Also tested on https://github.com/duduribeiro/testbug and it worked too (pushed the changes to use this new commit). 

REALLY thanks for this 

awesome work 🥇

jorgemanrubia committed:
Make sure the system to auto-filter params supports different initialization orders

This reworks the system to auto-filter params so that it works when encrypted
attributes are declared before the encryption configuration logic runs.

Co-authored-by: Cadu Ribeiro <mail@cadu.dev>

jorgemanrubia commented on activerecord/lib/active_record/railtie.rb:
```diff
@@ -357,11 +357,17 @@ class Railtie < Rails::Railtie # :nodoc:
-         **config.active_record.encryption
+      auto_filtered_parameters = ActiveRecord::Encryption::AutoFilteredParameters.new(app)
+
+      config.after_initialize do |app|
```
@rafaelfranca I went with using an `after_initialize` at the end because, even if it's discouraged, it can happen that apps include code that loads ActiveRecord before a given initializer runs, which was resulting in situations where AR encryption got initialized before certain properties had been processed.

maximerety replied:
I've created an issue and PR following that change:

* https://github.com/rails/rails/issues/50604
* https://github.com/rails/rails/pull/50606

I believe this change breaks compatibility with eager loading mode, in which AR models are loaded before the `after_initialize` hook has a chance to set up ARE configs. As a result, ARE's `encrypts` can't access configs that should be already ready but are not.

I'd like to hear your views on this issue.

Thanks.

jorgemanrubia commented:
@eileencodes this is good to merge from my side

trevorturk commented:
Apologies for commenting on an older PR here, but I've been stumbling over some issues with encryption while upgrading from 7.0.8 -> 7.1.3 and have been reading through old issues and PRs and noticed this comment: https://github.com/rails/rails/issues/48204#issuecomment-1598373096

> My first idea is adding an option to support decrypting data with both digest classes, so that we offer an upgrade path where things don't break for existing users.

...which would improve the situation w/r/t potentially having to downgrade (or slowly rolling out an upgrade with both versions running at the same time) re: https://github.com/rails/rails/issues/50226. I'm curious if that turned out to be more difficult than adding the `support_sha1_for_non_deterministic_encryption` option? (Of course, if we wanted to consider a new config option, it would need to be back ported to 7.0.x and documented, which may not be worth the effort.)

jorgemanrubia commented:
@trevorturk That's what the option does. When you toggle it on, it will support both digest algorithms for decrypting existing data. Isn't it working for your case?

trevorturk commented:
@jorgemanrubia I'm sorry, my mistake! I was looking for the config option in Rails 7.0 but now I see it was added in 7.1 🫠👍
